### PR TITLE
Service Worker

### PR DESCRIPTION
**Pull request by nathsimpson**

### CODE REVIEW

File: code/assets/js/script.min.js [hunk 1]
```diff
@@ -2,26 +2,24 @@
 function a(a,b){if(b){var c=!a.getAttribute("viewBox")&&b.getAttribute("viewBox"),d=document.createDocumentFragment(),e=b.cloneNode(!0);for(c&&a.setAttribute("viewBox",c);e.childNodes.length;)d.appendChild(e.firstChild);a.appendChild(d)}}function b(b){b.onreadystatechange=function(){if(4===b.readyState){var c=document.createElement("x");c.innerHTML=b.responseText,b.s.splice(0).map(function(b){a(b[0],c.querySelector("#"+b[1].replace(/(\W)/g,"\\$1")))})}},b.onreadystatechange()}function c(c){function d(){for(var c;c=e[0];){var j=c.parentNode;if(j&&/svg/i.test(j.nodeName)){var k=c.getAttribute("xlink:href");if(f&&(!g||g(k,j,c))){var l=k.split("#"),m=l[0],n=l[1];if(j.removeChild(c),m.length){var o=i[m]=i[m]||new XMLHttpRequest;o.s||(o.s=[],o.open("GET",m),o.send()),o.s.push([j,n]),b(o)}else a(j,document.getElementById(n))}}}h(d,17)}c=c||{};var e=document.getElementsByTagName("use"),f="shim"in c?c.shim:/\bEdge\/12\b|\bTrident\/[567]\b|\bVersion\/7.0 Safari\b/.test(navigator.userAgent)||(navigator.userAgent.match(/AppleWebKit\/(\d+)/)||[])[1]<537,g=c.validate,h=window.requestAnimationFrame||setTimeout,i={};f&&d()}return c});
 svg4everybody();
 
+/*
```
Comment (dominikwilkowski) [inline code/assets/js/script.min.js]: Need to reenable google analytics here

File: code/assets/js/script.min.js [hunk 2]
```diff
@@ -2,26 +2,24 @@
...
 window.dataLayer = window.dataLayer || [];
 function gtag(){dataLayer.push(arguments);}
 gtag('js', new Date());
 
 gtag('config', 'UA-23571257-4');
+*/
 
+async function init(){
```
Comment (dominikwilkowski) [inline code/assets/js/script.min.js]: We need to test if service workers are supported here first.

File: code/assets/js/script.min.js [hunk 3]
```diff
@@ -2,26 +2,24 @@
...
+	registration = await navigator.serviceWorker.register('/sw.min.js')
```
Comment (dominikwilkowski) [inline code/assets/js/script.min.js]: `try` `catch` missing

File: code/assets/js/script.min.js [hunk 4]
```diff
@@ -2,26 +2,24 @@
...
+	registration.onupdatefound = () => {
+		console.log("Installing new service worker");
+		registration.installing.onstatechange = function(){
+			console.log(this.state);
```
Comment (dominikwilkowski) [inline code/assets/js/script.min.js]: needs to be removed for production

File: code/assets/js/script.min.js [hunk 5]
```diff
@@ -2,26 +2,24 @@
...
-			console.info('= Service Worker Registered =');
-	});
-
-	navigator
-		.serviceWorker
-		.ready
-		.then( function( registration ) {
-			console.info('=    Service Worker Ready   =');
-	});
+	init();
```
Comment (dominikwilkowski) [inline code/assets/js/script.min.js]: A self calling function might be better or do the promise chain as I had it before

File: code/assets/js/sw.template.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
+self.oninstall = evt => {
+	evt.waitUntil(
+		caches.open($name).then(cache => {
```
Comment (dominikwilkowski) [inline code/assets/js/sw.template.js]: Can we make the template variables more visible so the script fails if it runs without templating? Something like `<<NAME>>` or `{{NAME}}` or something.
So this line would look like:
```js
caches.open(<<NAME>>).then(cache => {
```

File: code/assets/js/sw.template.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+			return cache.addAll($filesToCache);
+		})
+	);
+
+	// replaces the old SW automatically
+	self.skipWaiting();
+};
+
+self.onactivate = evt => {
```
Comment (dominikwilkowski) [inline code/assets/js/sw.template.js]: I'm a big proponent of naming variables in a human way. What does `evt` stand for? Should we name it a real word?

File: code/helper/sw.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
+const fs = require("fs");
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: single quotes please

File: code/helper/sw.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+const path = require("path");
+const PKG = require("../../package.json");
+const NAME = `"design-system-meetup-${PKG.version}"`;
+
+let filesToCache = ['"/"']
+
+function read(dir) {
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: Definitely needs a better function name

File: code/helper/sw.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	fs.readdirSync(dir).forEach(child => {
+		// if child is a file
+		if(path.extname(child)){
+			let item = `${dir.replace('docs', '')}/${child.replace('/index.html', '')}`;
+			if(item !== ''){filesToCache.push(`"${item}"`)}
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: you just assigned `item` with a string. This condition can never be false

File: code/helper/sw.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+		//otherwise it's a folder
+		}else if(!child.includes('CNAME')){
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: what about `.` files?

File: code/helper/sw.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+			read(`${dir}/${child}`);
+		}
+	})
+}
+
+read('docs');
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: Have a look at this function:
https://github.com/cuttlebelle/cuttlebelle/blob/master/src/site.js?ts=2#L37
There is a function that crawls a file system and returns all folders/files.

File: code/helper/sw.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+fs.readFile('code/assets/js/sw.template.js', 'utf8', function(err, template) {
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: You put the above thing into a function and this is executed directly with a hardcoded value. Let's check together how we might make this a bit more structured.

File: code/helper/sw.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
+const fs = require("fs");
+const path = require("path");
+const PKG = require("../../package.json");
+const NAME = `"design-system-meetup-${PKG.version}"`;
+
+let filesToCache = ['"/"']
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: You don't need the quotes. Use `JSON.stringify` or something else when replacing

File: code/helper/sw.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+read('docs');
+
+fs.readFile('code/assets/js/sw.template.js', 'utf8', function(err, template) {
+	//console.log('Service worker will cache these files...', filesToCache);
+	let file = template
+		.replace(/\$name/g, NAME)
+		.replace(/\$filesToCache/g, filesToCache)
+		.replace(/^\s*[\r\n]/gm, '');
+
+	fs.writeFile("docs/assets/js/sw.min.js", file, function(err) {
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: I would use an arrow function here and everywhere in this file

File: code/helper/sw.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+read('docs');
+
+fs.readFile('code/assets/js/sw.template.js', 'utf8', function(err, template) {
+	//console.log('Service worker will cache these files...', filesToCache);
+	let file = template
+		.replace(/\$name/g, NAME)
+		.replace(/\$filesToCache/g, filesToCache)
+		.replace(/^\s*[\r\n]/gm, '');
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: your way of minifying? Leaves a lot of white space still. Worth using an actual minifier?

File: code/layout/helper.js [hunk 17]
```diff
@@ -29,3 +30,14 @@ export const MakeDateTime = ( date, time ) => {
 		.hour( parseInt( hour ) + 12 )
 		.minute( minute );
 };
+
+export const RelativeURL = ( URL, ID ) => {
+	let relative = `${ Path.posix.relative( `/${ ID === 'index' ? '.' : ID }/`, `/${ URL }` ) }`;
+	console.log(`/${ ID === 'index' ? '.' : ID }/`, ' ', `/${ URL }`, ' --> ' , relative);
```
Comment (dominikwilkowski) [inline code/layout/helper.js]: definitely don't need that console log anymore

File: code/helper/sw.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+		}else if(!child.includes('CNAME')){
+			read(`${dir}/${child}`);
+		}
+	})
+}
+
+read('docs');
+
+fs.readFile('code/assets/js/sw.template.js', 'utf8', function(err, template) {
+	//console.log('Service worker will cache these files...', filesToCache);
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: No need for console logs in the git history

File: code/layout/page.js [hunk 19]
```diff
@@ -38,7 +38,7 @@ const Page = ({
 
 				{ footer }
 
-				<Scripts _ID={ _ID } _relativeURL={ _relativeURL }/>
+				<Scripts _ID={ _ID } _relativeURL={ RelativeURL }/>
```
Comment (dominikwilkowski) [inline code/layout/page.js:41]: since we removed this prop in `script.js` we don't need this anymore

File: code/layout/page.js [hunk 20]
```diff
@@ -15,7 +15,7 @@ const Page = ({
 }) => {
 	return (
 		<html>
-			<Head _ID={ _ID } _relativeURL={ _relativeURL } pagetitle={ pagetitle }/>
+			<Head _ID={ _ID } _relativeURL={ RelativeURL } pagetitle={ pagetitle }/>
```
Comment (dominikwilkowski) [inline code/layout/page.js:18]: No need for this prop anymore since Head is now importing it directly

File: code/layout/event.js [hunk 21]
```diff
@@ -27,7 +26,7 @@ const Event = ({
 	footer
 }) => (
 	<html>
-		<Head _ID={ _ID } _relativeURL={ _relativeURL } pagetitle={ pagetitle }/>
+		<Head _ID={ _ID } _relativeURL={ RelativeURL } pagetitle={ pagetitle }/>
```
Comment (dominikwilkowski) [inline code/layout/event.js]: no need

File: code/layout/event.js [hunk 22]
```diff
@@ -60,7 +59,7 @@ const Event = ({
 
 						<EventDetails
 							_parseMD={ _parseMD }
-							_relativeURL={ _relativeURL }
```
Comment (dominikwilkowski) [inline code/layout/event.js]: no need

File: code/layout/event.js [hunk 23]
```diff
@@ -78,7 +77,7 @@ const Event = ({
 
 			{ footer }
 
-			<Scripts _ID={ _ID } _relativeURL={ _relativeURL }/>
+			<Scripts _ID={ _ID } _relativeURL={ RelativeURL }/>
```
Comment (dominikwilkowski) [inline code/layout/event.js]: can go

File: code/layout/homepage.js [hunk 24]
```diff
@@ -28,7 +27,7 @@ const Homepage = ({
 
 	return (
 		<html>
-			<Head _ID={ _ID } _relativeURL={ _relativeURL } pagetitle={ pagetitle }/>
+			<Head _ID={ _ID } _relativeURL={ RelativeURL } pagetitle={ pagetitle }/>
```
Comment (dominikwilkowski) [inline code/layout/homepage.js:31]: can go

File: code/layout/homepage.js [hunk 25]
```diff
@@ -59,7 +58,7 @@ const Homepage = ({
 
 						<EventDetails
 							_parseMD={ _parseMD }
-							_relativeURL={ _relativeURL }
+							_relativeURL={ RelativeURL }
```
Comment (dominikwilkowski) [inline code/layout/homepage.js:62]: can go

File: code/layout/homepage.js [hunk 26]
```diff
@@ -122,7 +121,7 @@ const Homepage = ({
 
 				{ footer }
 
-				<Scripts _ID={ _ID } _relativeURL={ _relativeURL }/>
+				<Scripts _ID={ _ID } _relativeURL={ RelativeURL }/>
```
Comment (dominikwilkowski) [inline code/layout/homepage.js:125]: can go

File: content/index/index.yml [hunk 27]
```diff
@@ -7,7 +7,7 @@ header:
 next:
   pagetitle: Design System meetup Sydney v6.0.0
   version: v6.0.0
-  city: Sydney
+  city: Sydney CBD
```
Comment (dominikwilkowski) [inline content/index/index.yml]: why the change?

File: docs/assets/js/sw.min.js [hunk 28]
```diff
@@ -0,0 +1,42 @@
+self.oninstall = evt => {
+	evt.waitUntil(
+		caches.open("design-system-meetup-1.0.1").then(cache => {
+			return cache.addAll("/","/assets/css/site.min.css","/assets/favicons/android-chrome-192x192.png","/assets/favicons/android-chrome-512x512.png","/assets/favicons/apple-touch-icon.png","/assets/favicons/browserconfig.xml","/assets/favicons/favicon-16x16.png","/assets/favicons/favicon-32x32.png","/assets/favicons/favicon.ico","/assets/favicons/mstile-150x150.png","/assets/favicons/safari-pinned-tab.svg","/assets/favicons/site.webmanifest","/assets/img/atlassian.png","/assets/img/dta.png","/assets/img/expert360.png","/assets/img/invision.png","/assets/img/og-image.jpg","/assets/img/thinkmill.png","/assets/img/tomwalker.png","/assets/img/zip.png","/assets/js/script.min.js","/assets/js/sw.template.js","/assets/svg/sprite.svg","/code-of-conduct/index.html","/index.html","/sw.min.js","/v1.0.0","/v2.0.0","/v3.0.0","/v4.0.0","/v5.0.0");
```
Comment (dominikwilkowski) [inline docs/assets/js/sw.min.js]: need to exclude the service worker file

Comment (dominikwilkowski) [reply docs/assets/js/sw.min.js]: Also don't need the template

File: docs/assets/js/sw.template.js [hunk 29]
```diff
@@ -0,0 +1,51 @@
+self.oninstall = evt => {
```
Comment (dominikwilkowski) [inline docs/assets/js/sw.template.js]: This should no end up in the docs folder. Needs to be either git ignored or handled differently

File: docs/sw.min.js [hunk 30]
```diff
@@ -1,17 +1,78 @@
...
-});
-
-self
-	.addEventListener( 'activate', function( e ) {
-		self
-			.registration
-			.unregister()
-			.then( function() {
-				return self.clients.matchAll();
+const VERSION = '0.1.1';
```
Comment (dominikwilkowski) [inline docs/sw.min.js]: What's with this file? Now we have two SW?

File: yarn.lock [hunk 31]
```diff
@@ -5,190 +5,230 @@
 "@mrmlnc/readdir-enhanced@^2.2.1":
   version "2.2.1"
   resolved "https://registry.yarnpkg.com/@mrmlnc/readdir-enhanced/-/readdir-enhanced-2.2.1.tgz#524af240d1a360527b730475ecfa1344aa540dde"
+  integrity sha512-bPHp6Ji8b41szTOcaP63VlnbbO5Ny6dwAATtY6JTjh5N2OLrb5Qk/Th5cRkRQhkWCt+EJsYrNB0MiL+Gpn6e3g==
```
Comment (dominikwilkowski) [inline yarn.lock:125]: Let's not make unnecessary changes to the lock file. Please revert

File: code/helper/sw.js (hunk 15 above)
Comment (dominikwilkowski) [reply code/helper/sw.js]: That;s also the wrong place to write it out to. Service workers need to be in the root of the website. `docs/sw.min.js`

File: code/assets/js/script.min.js (hunk 3 above)
Comment (dominikwilkowski) [reply code/assets/js/script.min.js]: And you moved the service worker to another folder. This path is right but it now points to the old SW

**Review by dominikwilkowski — CHANGES_REQUESTED** [review]
Hey @nathansimpsondesign 
Thank you for all this work. Looks very good and is close. I added some notes here and there. Let's go over them together.
Have a good weekend.

File: content/index/index.yml (hunk 27 above)
Comment (nathsimpson) [reply content/index/index.yml]: resolved

File: code/assets/js/sw.template.js [hunk 32]
```diff
@@ -0,0 +1,50 @@
+self.oninstall = event => {
+	event.waitUntil(
+		caches.open({{NAME}}).then(cache => {
+			return cache.addAll([$filesToCache]);
```
Comment (dominikwilkowski) [inline code/assets/js/sw.template.js]: should be two curlies
```suggestion
			return cache.addAll([{{filesToCache}}]);
```

File: code/helper/sw.js [hunk 33]
```diff
@@ -0,0 +1,61 @@
...
+	return files;
+};
+
+const getTemplateText = (NAME, filesToCache) => {
+	return new Promise((resolve, reject) => {
+		try {
+			fs.readFile('code/assets/js/sw.template.js', 'utf8', (err, template) => {
+				let file = template
+					.replace(/\{\{NAME\}\}/g, NAME)
+					.replace(/\$filesToCache/g, filesToCache)
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: fix me

File: code/helper/sw.js [hunk 34]
```diff
@@ -0,0 +1,61 @@
...
+		}
+		);
+	return files;
+};
+
+const getTemplateText = (NAME, filesToCache) => {
+	return new Promise((resolve, reject) => {
+		try {
+			fs.readFile('code/assets/js/sw.template.js', 'utf8', (err, template) => {
+				let file = template
```
Comment (dominikwilkowski) [inline code/helper/sw.js]:
```suggestion
				const file = template
```

File: code/helper/sw.js [hunk 35]
```diff
@@ -0,0 +1,61 @@
...
+			}
+		}
+		);
+	return files;
+};
+
+const getTemplateText = (NAME, filesToCache) => {
+	return new Promise((resolve, reject) => {
+		try {
+			fs.readFile('code/assets/js/sw.template.js', 'utf8', (err, template) => {
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: should probably use `readFileSync`

File: code/layout/eventDetails.js [hunk 36]
```diff
@@ -10,9 +10,9 @@ import React, { Fragment } from 'react';
  * The Event detail component
  */
 const EventDetails = ({
-	_relativeURL,
 	_ID,
 	_parseMD,
+	_relativeURL,
```
Comment (dominikwilkowski) [inline code/layout/eventDetails.js]: should us the helpers `relativeURL`

**Review by dominikwilkowski — CHANGES_REQUESTED** [review]
Move to helper `relativeURL` and remove promise in `sw.js`

File: code/helper/sw.js [hunk 37]
```diff
@@ -0,0 +1,61 @@
...
+				let file = template
+					.replace(/\{\{NAME\}\}/g, NAME)
+					.replace(/\$filesToCache/g, filesToCache)
+					.replace(/^\s*[\r\n]/gm, '');
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: maybe use uglify2?

File: code/assets/js/sw.template.js [hunk 38]
```diff
@@ -0,0 +1,50 @@
+self.oninstall = event => {
+	event.waitUntil(
+		caches.open({{NAME}}).then(cache => {
+			return cache.addAll([{{filesToCache}}]);
```
Comment (dominikwilkowski) [inline code/assets/js/sw.template.js]: Let's do the same as with `{{NAME}}`.

```suggestion
			return cache.addAll([{{FILES}}]);
```

**Review by dominikwilkowski — CHANGES_REQUESTED** [review]

File: code/helper/sw.js [hunk 39]
```diff
@@ -34,28 +35,16 @@ const listFilesInFolder = (dir, root) => {
...
+const sw_file = getTemplateText(NAME, filesToCache);
 
-getTemplateText(NAME, filesToCache)
-	.then(file => {
-		fs.writeFile('docs/sw.min.js', file, error => {
-			if (error) console.log(error);
-		});
-	})
-	.catch(error => console.log(error));
+fs.writeFileSync('docs/sw.min.js', sw_file, error => {if (error) console.log(error);});
```
Comment (dominikwilkowski) [inline code/helper/sw.js]: The third argument for `writeFileSync` is `options`
https://nodejs.org/api/fs.html#fs_fs_writefilesync_file_data_options

However `fs_writefile` does have a FORTH argument for callback functions.
https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback

It does not take a function

File: code/helper/sw.js [hunk 40]
```diff
@@ -0,0 +1,50 @@
...
+	return files;
+};
+
+const getTemplateText = (NAME, filesToCache) => {
+	let template = fs
+		.readFileSync('code/assets/js/sw.template.js', 'utf8')
+		.replace(/\{\{NAME\}\}/g, NAME)
+		.replace(/\{\{FILES\}\}/g, filesToCache);
+
+	let minified = minify(template);
```
Comment (dominikwilkowski) [inline code/helper/sw.js]:
```suggestion
	const minified = minify(template);
```

**Review by dominikwilkowski — CHANGES_REQUESTED** [review]
So close! Good work!

**Pull request merged**